### ISSUE TITLE
Limiting the number of temporary file for a rule

### ISSUE THREAD

**Issue by jsgounot** [snakemake/snakemake#1896]
**Is your feature request related to a problem? Please describe.**
Similar to a lot of users, temporary files are often use to quickly remove big temporary files. In some cases, the system is not able to keep on disk all the temporary files at once. Since some versions ago, the scheduler has been modified to improve this situation but this is still not perfect. In some occasion, jobs are still executed in a breadth-first manner, even after using `resources` or `priority`. For example, this seems to happen if rule A generates a temporary file with 4 cores and rule B using temporary files from rule A needs 8 cores. The scheduler will favor rule A and all temporary files will be generated before being used.

**Describe the solution you'd like**
Add an optional argument to `temp`, which would force snakemake to keep at maximum x number of temporary files during the execution. If x number of temporary files are produced, snakemake will do any other rule or wait to have resource to run any other rule. If x is not given, current behavior can remain the same.

**Comment by dariober** [comment]
>  the system is not able to keep on disk all the temporary files at once.

Out of curiosity, what is the limiting factor? The number of files itself or their total size? If the latter, could [resources](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#standard-resources) (`disk_mb`, in particular) help?

**Comment by jsgounot** [comment]
The total size. Unless I misunderstood, `resources` limit the number of parallel executions of a rule but does not take into consideration the cumulative size of temporary files in its calculation. Funny enough, I've also an issue with the number of files I presented [here](https://github.com/snakemake/snakemake/issues/1877).

**Comment by dariober** [comment]
Double-check what I'm going to say here, maybe run some toy scripts... 

When a rule has `resources: disk_mb=1000` you tell snakemake that that rule requires 1000 Mb of disk space. So when you execute with `--resources mem_mb=2000`, snakemake will make sure that no more than two instances of that rule will run at the same time. This assumes that no other rule has been constrained with `resources: disk_mb=...`, otherwise snakemake will arrange things not to exceed a total of 2000 Mb.

This means that you have to tell snakemake how much disk space a rule needs (at most) but I think snakemake does not keep track of the file sizes a rule generates (which would be nice but I wonder if it is at all possible).

Similarly, I think you could have rules with `resources: number_of_files=100`. Running `snakemake --resources number_of_files=1000` you tell snakemake not to stay below 1000 files (i.e. no more than 10 of those rules at the same time).

**Comment by jsgounot** [comment]
Yeah I already tried that. The issue is not that I'm not able to limit the number of execution of a rule at the same time, but it's the fact that the generated temporary file is an output for another rule. Meaning that when the rule finishes, the disk space is still used and will remain unless the downstream rule(s) is executed (which might not be the case due to CPU constrain for example).
